packages/grafana-ui/src/components/Tags/Tag.tsx

```diff
@@ -22,19 +22,24 @@ export const Tag = forwardRef<HTMLElement, Props>(({ name, onClick, className, c
   const styles = getTagStyles(theme, name, colorIndex);
 
   const onTagClick = (event: React.MouseEvent<HTMLElement>) => {
-    if (onClick) {
-      onClick(name, event);
-    }
+    event.preventDefault();
+    event.stopPropagation();
+
+    onClick?.(name, event);
+  };
+
+  const commonProps = {
+    key: name,
+    className: cx(styles.wrapper, className, onClick && styles.hover),
+    ...rest,
   };
 
-  return (
-    <span
-      key={name}
-      ref={ref}
-      onClick={onTagClick}
-      className={cx(styles.wrapper, className, onClick && styles.hover)}
-      {...rest}
-    >
+  return onClick ? (
+    <button {...commonProps} onClick={onTagClick} ref={ref as React.ForwardedRef<HTMLButtonElement>}>
```

Comment:
hmmm... i don't like this `as HTMLButtonElement`. It prevents consumers from actually using a HTMLButtonElement ref.

I'll investigate options. Maybe `aria-role=button` as a last resort?

Reply:
One idea would be instead of`forwardRef`, to use an `innerRef` prop, and the props can be a union of two interfaces with the two different ref types

```typescript
interface BaseProps { ... }
interface SpanTagProps  extends BaseProps {
  innerRef: RefObject<HTMLElement>;
}
interface ButtonTagProps  extends BaseProps {
  onClick: () => void; // eg
  innerRef: RefObject<HTMLButtonElement>;
}

export type Props = SpanTagProps | ButtonTagProps;

export default const Tag = (props: Props) => {
  if ("onClick" in props) {
    <button onClick={props.onClick} ref={props.innerRef} /?
  } else {
    ...
  }
}
```

Reply:
The only unfortunate thing here is that it'd be a breaking change, which I'd rather avoid :(

Reply:
`ref={ref as React.ForwardedRef<HTMLButtonElement>}` is already a breaking change as _at runtime_ you're passing a different type of ref up to the consumer.

So it's either "have a subtle, hidden breaking change with incorrect types" or "have an explicit breaking change with correct types". By typing this correctly, we will know where the actual breakages are.

Reply:
I just think it would be a shame to have an accessibility fix be delayed until 9.0 when the interface itself wouldn't be broken. Consumers of `Tag` currently receive an `HTMLElement` ref, which is what they would still get, even if it's now a more specific kind of `HTMLElement`. What do you imagine might be broken by doing things this way?

Reply:
Hmmm yes, I understand what you're saying - it's not that much of a breaking change as HTMLButtonElement is still a superclass of HTMLElement.

Short-medium term we would be looking to remove all the `as` usage, so it would be good to sort this out some time :)